### PR TITLE
Refactoring changes needed for us to implement empty assembly output

### PR DESCRIPTION
**Pull request by mrvoorhe**
In UnityLinker we are adding an option that causes assemblies that would normally be deleted
to instead be written out as empty assemblies.  This is an ability we would like for our build system.

This ability is not desired in monolinker, but I needed to move a few things around in order to implement this on our end.

* Expose resolve methods on `ResultChecker`

* Expose `RemoveAssembly` and make it virtual

* Split `SweepAssembly`.  I'd like to use the logic that does the actual removal of everything when clearing out an assembly that would normally be deleted.
This way all of the `ElementRemoved` methods are called, which we use for logging purposes.  By moving the switch part of `SweepAssembly` out into a new method called `ProcessAssembly`,
I can call `SweepAssembly` from our overridden `RemoveAssembly`

* Don't mark top level attrs until the module is marked.  Normally you would never notice this was happening because
the entire assembly would be deleted anyways.  But when you want `SweepAssembly` to remove as much as possible for an assembly that would normally be Deleted it is helpful if `CustomAttribute`
in that assembly have not been marked

### CODE REVIEW

File: linker/Linker.Steps/MarkStep.cs [hunk 1]
```diff
@@ -480,6 +480,11 @@ protected virtual bool ShouldMarkTopLevelCustomAttribute (AttributeProviderPair
 			// then surely nothing is using this attribute and there is no need to mark it
 			if (!Annotations.IsMarked (resolvedConstructor.Module) && !Annotations.IsMarked (ca.AttributeType))
 				return false;
+
+			// If module the custom attribute appears in has not been marked yet, then don't mark the attribute
+			// i.e Don't mark custom attributes on an assembly until the assembly has been marked
+			if (!Annotations.IsMarked (ca.AttributeType.Module))
```
[the remark this answers is not in the document]
Comment (mrvoorhe) [reply linker/Linker.Steps/MarkStep.cs]: The check before this is

`if (!Annotations.IsMarked (resolvedConstructor.Module) && !Annotations.IsMarked (ca.AttributeType))`

I.e. If the resolved module is not marked and the attribute TypeReference is not marked.

Imagine `[assembly: Debuggable]` in an assembly that we are going to strip away because nothing is used.

`Annotations.IsMarked (resolvedConstructor.Module) ` is going to be true because `resolvedConstructor.Module` is mscorlib.

With this new if `if (!Annotations.IsMarked (ca.AttributeType.Module))` , `Annotations.IsMarked (ca.AttributeType.Module)` is going to be false because `ca.AttributeType.Module` is our assembly that we can strip away and it has not been marked

Maybe, the two if's can be combine.  I didn't try and think about it for a few reasons

* This is kind of delicate edge case stuff and I don't want to introduce a regression

* I can't (easily) write a test in the upstream linker to cover this new if I'm adding because the assembly is going to be deleted anyways.  We'd need this "no delete" mode I'm adding to our UnityLinker in order to test this.  As a result, I'd rather the logic be it's own dedicated if rather than mixed into some other if where it's purpose might be less clear.

[the remark this answers is not in the document]
Comment (marek-safar) [reply linker/Linker.Steps/MarkStep.cs]: I don't like adding yet another combination without understanding existing logic.

We already have a case which does same thing here https://github.com/mono/linker/blob/44f3b1e740b7156b7109d04a9f58a0473c78a466/linker/Linker.Steps/MarkStep.cs#L465 but under KeepUsedAttributeTypesOnly. This kind of removes the condition.

Secondly, I am not sure why do you have `.Module` there. It basically blocks any attributes declared in mscorlib to be ever stripped when referenced.

[the remark this answers is not in the document]
Comment (mrvoorhe) [reply linker/Linker.Steps/MarkStep.cs]: The goal of the new if I'm adding is to not mark `CustomAttribute` instances on the Assembly when the Assembly is going to be deleted regardless of whether `KeepUsedAttributeTypesOnly` is enabled or not.

The goal of the if you pointed out is to not mark `CustomAttribute` instances until the TypeDefinition of the AttributeType has been marked.

> Secondly, I am not sure why do you have .Module there. It basically blocks any attributes declared in mscorlib to be ever stripped when referenced.

I don't follow what your concern is.  I added a change set to this PR to support asserting attributes in other assemblies and I added new tests to assert what happens with attributes defined in mscorlib.  Nothing seems wrong..
https://github.com/mono/linker/pull/360/commits/7a08b2e51b466bae6e4ff3accd9d91497553056f

**Comment by mrvoorhe** [comment]
@marek-safar I dropped the assembly attribute behavior change from this PR.  This PR is now refactoring only.

I will open a separate PR for the assembly attribute change and we can continue the discussion there.
